// @wiitate/eslint-plugin/config/core.js
module.exports = {
  parserOptions: {
    ecmaVersion: 'latest',
  },
  env: {
    node: true,
    es6: true,
    jest: true,
  },
  plugins: ['unicorn', 'prettier', 'import'],
  extends: ['eslint:recommended', 'prettier', 'plugin:import/recommended'],
  rules: {
    /** Possible Errors */
    'no-await-in-loop': 'warn',
    'no-console': ['warn', { allow: ['warn', 'error'] }],
    'no-undef': 'off',
    'no-template-curly-in-string': 'warn',
    'require-atomic-updates': 'off',
    'no-extra-boolean-cast': 'off',
    'no-dupe-else-if': 'off',

    /** Best Practices */
    'class-methods-use-this': 'off',
    'consistent-return': 'warn',
    'default-case': 'error',
    'default-param-last': 'off',
    'no-alert': 'error',
    'no-caller': 'error',
    'no-div-regex': 'warn',
    'no-empty-function': 'error',
    'no-eval': 'error',
    'no-floating-decimal': 'error',
    'no-multi-spaces': 'error',
    'no-new': 'error',
    'no-new-func': 'error',
    'no-proto': 'error',
    'no-self-compare': 'error',
    'no-useless-call': 'error',
    'require-await': 'error',
    'no-unused-expressions': 'off',

    /** Variables */
    'no-unused-vars': 'error',

    /** Stylistic Issues */
    'array-bracket-spacing': ['error', 'never'],
    'block-spacing': 'error',
    camelcase: 'error',
    'comma-spacing': ['error', { before: false, after: true }],
    'computed-property-spacing': ['error', 'never'],
    'func-call-spacing': ['error', 'never'],
    'func-names': ['error', 'always'],
    'jsx-quotes': ['error', 'prefer-double'],
    'key-spacing': ['error', { mode: 'strict' }],
    'new-parens': 'error',
    'no-bitwise': 'error',
    'no-multiple-empty-lines': ['error', { max: 1, maxEOF: 1 }],
    'no-new-object': 'error',
    'no-trailing-spaces': 'error',
    'no-whitespace-before-property': 'error',
    'object-curly-spacing': ['error', 'always'],
    'semi-spacing': ['error', { before: false, after: true }],
    'semi-style': ['error', 'last'],
    'space-before-blocks': 'error',
    'space-before-function-paren': [
      'error',
      { anonymous: 'never', named: 'never', asyncArrow: 'always' },
    ],
    'space-in-parens': ['error', 'never'],
    'space-infix-ops': 'error',
    'space-unary-ops': 'error',
    'spaced-comment': ['warn', 'always'],
    'switch-colon-spacing': ['error', { after: true, before: false }],
    'wrap-regex': 'off',

    /** ECMAScript 6 */
    'arrow-parens': 'error',
    'arrow-spacing': 'error',
    'no-confusing-arrow': 'off',
    'no-duplicate-imports': 'error',
    'no-useless-constructor': 'error',
    'no-var': 'error',
    'prefer-const': 'error',
    'prefer-arrow-callback': ['error', { allowNamedFunctions: false }],
    'template-curly-spacing': ['error', 'never'],

    /** Extensions */
    'import/no-unresolved': 'off',
    'unicorn/filename-case': [
      'error',
      {
        cases: {
          camelCase: true,
          kebabCase: true,
        },
      },
    ],
  },
};
